File: librarymanager/librarymanager-be/src/main/java/ru/spring5/elena/librarymanager/services/OrderService.java
package ru.spring5.elena.librarymanager.services;

import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import ru.spring5.elena.librarymanager.api.dto.Order;
import ru.spring5.elena.librarymanager.domain.entity.EditionEntity;
import ru.spring5.elena.librarymanager.domain.entity.ReaderEntity;
import ru.spring5.elena.librarymanager.domain.repository.EditionRepository;
import ru.spring5.elena.librarymanager.domain.repository.OrderRepository;
import ru.spring5.elena.librarymanager.domain.repository.ReaderRepository;
import ru.spring5.elena.librarymanager.mapper.OrderMapper;

import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;


@RequiredArgsConstructor
@Service
public class OrderService {
    OrderRepository orderRepository;
    ReaderRepository readerRepository;
    EditionRepository editionRepository;

    @Transactional(readOnly = true)
    public Page<Order> findAllByPage(Pageable pageable) {
        return orderRepository.findAll(pageable)
                .map(OrderMapper::map);
    }

    @Transactional(readOnly = true)
    public List<Order> findAll() {
        return orderRepository.findAll()
                .stream()
                .map(OrderMapper::map)
                .collect(Collectors.toList());
    }

    @Transactional(readOnly = true)
    public Order findOrderById(Long orderId) {
        return orderRepository.findById(orderId)
                .map(OrderMapper::map)
                .orElse(null);
    }

    @Transactional
    public Order save(Long orderId,  Order dto) {
        
    	ReaderEntity readerEntity = readerRepository.findById(dto.getReaderId()).get();
        EditionEntity editionEntity = editionRepository.findById(dto.getEditionId()).get();
        
    	return orderRepository.findById(orderId)
                .map(entity -> OrderMapper.map(dto, entity, readerEntity, editionEntity))
                .map(orderRepository::saveAndFlush)
                .map(OrderMapper::map)
                .orElse(null);

    }

    @Transactional
    public Order create(Order dto) {
    	
    	ReaderEntity readerEntity = readerRepository.findById(dto.getReaderId()).get();
        EditionEntity editionEntity = editionRepository.findById(dto.getEditionId()).get();
    	
        return   Optional.of(dto)
                .map(order -> OrderMapper.map(order, readerEntity, editionEntity))
                .map(orderRepository::saveAndFlush)
                .map(OrderMapper::map)
                .orElse(null);

    }

}

File: librarymanager-be/src/main/java/ru/elenacherev/librarymanager/services/OrderService.kt
package ru.elenacherev.librarymanager.services

import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import ru.elenacherev.librarymanager.api.dto.Order
import ru.elenacherev.librarymanager.domain.entity.EditionEntity
import ru.elenacherev.librarymanager.domain.entity.OrderEntity
import ru.elenacherev.librarymanager.domain.entity.ReaderEntity
import ru.elenacherev.librarymanager.domain.repository.EditionRepository
import ru.elenacherev.librarymanager.domain.repository.OrderRepository
import ru.elenacherev.librarymanager.domain.repository.ReaderRepository
import ru.elenacherev.librarymanager.mapper.map
import java.util.*

@Service
class OrderService(
    val orderRepository: OrderRepository,
    val readerRepository: ReaderRepository,
    val editionRepository: EditionRepository
) {
    @Transactional(readOnly = true)
    fun findAllByPage(pageable: Pageable): Page<Order> {
        return orderRepository
            .findAll(pageable)
            .map(OrderEntity::map)
    }

    @Transactional(readOnly = true)
    fun findAll(): List<Order> {
        return orderRepository
            .findAll()
            .map(OrderEntity::map)
    }

    @Transactional(readOnly = true)
    fun findOrderById(orderId: UUID): Order {
        return orderRepository
            .findById(orderId)
            .map(OrderEntity::map)
            .orElse(null)
    }

    @Transactional
    fun save(orderId: UUID, dto: Order): Order {
        val readerEntity: ReaderEntity = readerRepository.findById(dto.readerId).get()
        val editionEntity: EditionEntity = editionRepository.findById(dto.editionId).get()
        return orderRepository.findById(orderId)
            .map { entity -> dto.map(entity, readerEntity, editionEntity) }
            .map { order: OrderEntity -> orderRepository.saveAndFlush(order) }
            .map(OrderEntity::map)
            .orElse(null)
    }

    @Transactional
    fun create(dto: Order): Order {
        val readerEntity: ReaderEntity = readerRepository.findById(dto.readerId).get()
        val editionEntity: EditionEntity = editionRepository.findById(dto.editionId).get()
        return Optional.of(dto)
            .map { order -> order.map(readerEntity, editionEntity) }
            .map { order: OrderEntity -> orderRepository.saveAndFlush(order) }
            .map(OrderEntity::map)
            .orElse(null)
    }
}